File: src/leetcode/fourth/Code_094_InorderTraversal.java
package leetcode.fourth;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

import leetcode.tool.TreeNode;

/**
 * 二叉树的中序遍历
 * 迭代
 * 
 * @since 2020-8-4 Tuesday 9:37 - 10:01
 */
public class Code_094_InorderTraversal {
    static List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        if (root != null) {
            Stack<TreeNode> stack = new Stack<>();
            stack.push(root);
            root = root.left;
            while (!stack.isEmpty()) {
                while (root != null) {
                    stack.push(root);
                    root = root.left;
                }
                root = stack.pop();
                res.add(root.val);
                root = root.right;
                if (root != null) {
                    stack.push(root);
                    root = root.left;
                }
            }
        }
        return res;
    }

    static List<Integer> _inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        if (root !=  null) {
            Stack<TreeNode> stack = new Stack<>();
            while (!stack.isEmpty() || root != null) {
                if (root != null) {
                    stack.push(root);
                    root = root.left;
                } else {
                    root = stack.pop();
                    res.add(root.val);
                    root = root.right;
                }
            }
        }
        return res;
    }

    public static void main(String[] args) {
        System.out.println(inorderTraversal(TreeNode.tree()));
        System.out.println(_inorderTraversal(TreeNode.tree()));
    }
}

File: src/cpp-leetcode/first/Code_094_InorderTraversal.cpp
#include <iostream>
#include <stack>
#include <vector>

#include "../Tree.h"
using std::vector;

/**
 * inorder traversal of binary tree
 * iterate
 *
 * @since 2020-9-22 Tuesday 10:34 - 10:51
 */
class Solution {
   public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        if (root != nullptr) {
            std::stack<TreeNode*> stack;
            stack.push(root);
            root = root->left;
            while (!stack.empty()) {
                while (root != nullptr) {
                    stack.push(root);
                    root = root->left;
                }
                TreeNode* node = stack.top();
                stack.pop();
                res.push_back(node->val);
                if (node->right != nullptr) {
                    root = node->right;
                    stack.push(root);
                    root = root->left;
                }
            }
        }
        return res;
    }
};

int main(int argc, char const* argv[]) {
    TreeNode* root = Tree::binaryTree();
    Solution s;
    vector<int> res = s.inorderTraversal(root);
    for (int i : res) {
        std::cout << i << " ";
    }
}
